### ISSUE TITLE
[Feature] Support resizing

### ISSUE THREAD

**Issue by kvaps** [mariadb-operator/mariadb-operator#262]
**Is your feature request related to a problem? Please describe.**

could not update db template on the fly:
```diff
@@ -40,7 +40,7 @@ spec:
   volumeClaimTemplate:
     resources:
       requests:
-        storage: 1Gi
+        storage: 10Gi
     accessModes:
       - ReadWriteOnce
```

```
Error: UPGRADE FAILED: cannot patch "grafana-db" with kind MariaDB: admission webhook "vmariadb.kb.io" denied the request: MariaDB.mariadb.mmontes.io "grafana-db" is invalid: VolumeClaimTemp
late: Invalid value: v1alpha1.VolumeClaimTemplate{PersistentVolumeClaimSpec:v1.PersistentVolumeClaimSpec{AccessModes:[]v1.PersistentVolumeAccessMode{"ReadWriteOnce"}, Selector:(*v1.LabelSele
ctor)(nil), Resources:v1.ResourceRequirements{Limits:v1.ResourceList(nil), Requests:v1.ResourceList{"storage":resource.Quantity{i:resource.int64Amount{value:10737418240, scale:0}, d:resource
.infDecAmount{Dec:(*inf.Dec)(nil)}, s:"10Gi", Format:"BinarySI"}}, Claims:[]v1.ResourceClaim(nil)}, VolumeName:"", StorageClassName:(*string)(nil), VolumeMode:(*v1.PersistentVolumeMode)(nil)
, DataSource:(*v1.TypedLocalObjectReference)(nil), DataSourceRef:(*v1.TypedObjectReference)(nil)}, Labels:map[string]string(nil), Annotations:map[string]string(nil)}: 'VolumeClaimTemplate' f
ield is inmutable
```

**Describe the solution you'd like**

- Allow modifying`volumeClaimTemplate.resources.requests.storage` size.
- Patch requested size for all existing PVCs manually by operator

**Describe alternatives you've considered**

**Additional context**

**Environment details**:
- Kubernetes version: v1.27.4
- mariadb-operator version: v5.3.0
- Install method: helm
- Install flavour: minimal, recommended or custom

**Comment by Guzzi** [comment]
Hi @mmontes11 ,
afaik, the storage SIZE does not need to be inmutable - actually, it is depending on the underlying storageprovider, if it is extendable - operator could simply reconcile, if not, it will throw respective error and won't be ablöe to reconcile it.
It is different with storageClass - this can't be changed without deleting and recreating the pvc - so if that is something to be able to change, indeed a new statefulset with storage would have to be created and ideally data transferred - but tbh I would not support that, as it is a pretty complicated process - so imho I would allow changing the size and properly handle errors in the operator and keep storageClass inmutable.

**Comment by kvaps** [comment]
The resizing opportunity for PVCs is controlled by: `allowVolumeExpansion: true` parameter in storage class.

> As an alternative, we could ditch the StatefulSet and manage the Pods and PVCs ourselves from the operator. This will be a substantial gain in flexibility but it will have a huge impact in the implementation, as we heavily rely on the StatefulSet construct. Thoughts?

To me this can be implemented in two phases:

**phase 1**: update and set size for existing PVCs, leave pvcTemplate in StatefulSet unchanged (must have)
**phase 2**: update StatefulSet pvcTemplate with `delete --cascade=orphan` and  create a new one. This will take effect only for newly created PVCs (to me this it is optional, and can be implemented sometimes later)

upstream ref https://github.com/kubernetes/kubernetes/issues/68737

**Comment by kvaps** [comment]
btw, cloudnativepg use:
```
  storage:
    resizeInUseVolumes: true
    size: 100Gi
```

https://cloudnative-pg.io/documentation/1.15/api_reference/#StorageConfiguration

**Comment by mmontes11** [comment]
> phase 1: reconcile and set size for existing PVCs, leave pvcTemplate in StatefulSet unchanged (must have)
phase 2: reconcile StatefulSet's pvcTemplate with delete --cascade=orphan and create a new one. This will take effect only for newly created PVCs (to me this it is optional, and can be implemented some time later)

I think a phased approach makes total sense. To implement the phase 1, and to keep `volumeClaimTemplate` inmutable, we could introduce the `storage` field,  same as CNPG:

```
  storage:
    resizeInUseVolumes: true
    size: 100Gi
```
This field would be mutable, and when provided, it would make the operator to update the existing PVCs. If the PVC `storageClass` does not allow expansion, we will set an error on the `MariaDB` resource and also emit an error `Event`.

Does this sounds reasonable? We will try to accomodate this in the `v0.0.23`

**Comment by kvaps** [comment]
Sounds good to me

**Comment by mmontes11** [comment]
This feature has been supported via the following PR:
- https://github.com/mariadb-operator/mariadb-operator/pull/407

It will be released in `v0.0.26`. Please take a look at the new `storage` field for making use of it:
- https://github.com/mariadb-operator/mariadb-operator/blob/main/docs/API_REFERENCE.md#storage
